snowflake/src/main/java/org/apache/iceberg/snowflake/JdbcSnowflakeClient.java

```diff
@@ -208,6 +219,7 @@ public List<SnowflakeIdentifier> listDatabases() {
                   queryHarness.query(
                       conn, "SHOW DATABASES IN ACCOUNT", DATABASE_RESULT_SET_HANDLER));
     } catch (SQLException e) {
+      tryMapSnowflakeExceptionToIcebergException(SnowflakeIdentifier.ofRoot(), e);
```

Comment:
reading through the catch clause here and seeing this,  one wouldn't expect `tryMapSnowflakeExceptionToIcebergException` to actually throw. I think it would be good to rewrite the code so that it's more obvious.

What about having `snowflakeExceptionToIcebergException` return an `Optional<RuntimeException>`? Then we could simply have  `throw snowflakeExceptionToIcebergException(SnowflakeIdentifier.ofRoot(), e).orElseGet(() -> new UncheckedSQLException(e, "Failed to list databases"));` here.